File: parallel_merge_sort.py
import threading
import time

THRESHOLD = 5  # Minimum size for parallel sort

def merge(array, left, mid, right):
    left_size = mid - left + 1
    right_size = right - mid

    left_array = array[left:left + left_size]
    right_array = array[mid + 1:mid + 1 + right_size]

    i = j = 0
    k = left

    while i < left_size and j < right_size:
        if left_array[i] <= right_array[j]:
            array[k] = left_array[i]
            i += 1
        else:
            array[k] = right_array[j]
            j += 1
        k += 1

    while i < left_size:
        array[k] = left_array[i]
        i += 1
        k += 1

    while j < right_size:
        array[k] = right_array[j]
        j += 1
        k += 1

def merge_sort_thread(params):
    array, left, right = params
    if left < right:
        mid = left + (right - left) // 2

        if (right - left) <= THRESHOLD:
            # Sequential sort if below threshold
            merge_sort_thread((array, left, mid))
            merge_sort_thread((array, mid + 1, right))
        else:
            # Parallel sort if above threshold
            left_thread = threading.Thread(target=merge_sort_thread, args=((array, left, mid),))
            right_thread = threading.Thread(target=merge_sort_thread, args=((array, mid + 1, right),))

            left_thread.start()
            right_thread.start()

            left_thread.join()
            right_thread.join()

        merge(array, left, mid, right)


shared_variable = 0
mutex = threading.Lock()
def thread_func(thread_id):
    global shared_variable
    mutex.acquire()
    shared_variable += 1
    mutex.release()


def main():

    file_path = "C:\\Users\\luisa\\Desktop\\SCS\\SCS_Project\\MS_input.txt"

    try:
        with open(file_path, 'r') as file:

            first_line = file.readline().strip()
            n = int(first_line.split(':')[1].strip())

            second_line = file.readline().strip()
            array = list(map(int, second_line.split(':')[1].strip().split()))

        if len(array) != n:
            raise ValueError("Mismatch between the number of elements and the size specified in the file.")

        # print(f"Number of elements: {n}")
        # print("Unsorted array:")
        # print(" ".join(map(str, array)))

        start_time = time.time()
        merge_sort_thread((array, 0, n - 1))
        end_time = time.time()

        execution_time = end_time - start_time
        #print(f"\nExecution time: {execution_time:.6f} seconds.")

        thread_creation_time = 0.0
        start_context_switch = time.time()
        threads = []

        for i in range(n):
            thread_id = i + 1

            thread_start_time = time.time()
            thread = threading.Thread(target=thread_func, args=(thread_id,))
            thread.start()
            threads.append(thread)
            thread_end_time = time.time()

            thread_creation_time += (thread_end_time - thread_start_time)

        for thread in threads:
            thread.join()

        end_context_switch = time.time()  # Record end time for context switch

        context_switch_time = end_context_switch - start_context_switch
        avg_th_creation_time = thread_creation_time / n

        out_file_path = "C:\\Users\\luisa\\Desktop\\SCS\\SCS_Project\\outputs.txt"
        try:
            with open(out_file_path, 'a') as file:
                file.write(f"\nParallel merge sort in Python results\n")
                file.write(f"\t(ms-py) Execution time: {execution_time:.6f} seconds.\n")
                file.write(f"\t(ms-py) Average thread creation time: {avg_th_creation_time:.6f} seconds.\n")
                file.write(f"\t(ms-py) Context switch time: {context_switch_time:.6f} seconds.\n")
        except Exception as e:
            print(f"Error writing to file: {e}")

        # print("\nSorted array:")
        # print(" ".join(map(str, array)))

    except FileNotFoundError:
        print(f"Error: File not found at {file_path}")
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()

File: ParallelMergeSort.java
package org.example;

import java.io.*;
import java.util.Locale;
import java.util.concurrent.locks.ReentrantLock;

public class ParallelMergeSort {
    private static final int THRESHOLD = 5;

    private static int sharedVariable = 0;
    private static final ReentrantLock mutex = new ReentrantLock();

    static class WorkerThread extends Thread {
        private final int threadId;

        public WorkerThread(int threadId) {
            this.threadId = threadId;
        }

        @Override
        public void run() {
            mutex.lock();
            try {
                sharedVariable++;
            } finally {
                mutex.unlock();
            }
        }
    }

    private static void merge(int[] array, int left, int mid, int right) {
        int leftSize = mid - left + 1;
        int rightSize = right - mid;

        int[] leftArray = new int[leftSize];
        int[] rightArray = new int[rightSize];

        System.arraycopy(array, left, leftArray, 0, leftSize);
        System.arraycopy(array, mid + 1, rightArray, 0, rightSize);

        int i = 0, j = 0, k = left;

        while (i < leftSize && j < rightSize) {
            if (leftArray[i] <= rightArray[j]) {
                array[k++] = leftArray[i++];
            } else {
                array[k++] = rightArray[j++];
            }
        }

        while (i < leftSize) {
            array[k++] = leftArray[i++];
        }

        while (j < rightSize) {
            array[k++] = rightArray[j++];
        }
    }

    // MergeSortTask thread class
    private static class MergeSortTask extends Thread {
        private int[] array;
        private int left;
        private int right;

        public MergeSortTask(int[] array, int left, int right) {
            this.array = array;
            this.left = left;
            this.right = right;
        }

        @Override
        public void run() {
            if (left < right) {
                int mid = left + (right - left) / 2;

                if (right - left <= THRESHOLD) {
                    mergeSort(array, left, mid);
                    mergeSort(array, mid + 1, right);
                } else {
                    MergeSortTask leftTask = new MergeSortTask(array, left, mid);
                    MergeSortTask rightTask = new MergeSortTask(array, mid + 1, right);

                    leftTask.start();
                    rightTask.start();

                    try {
                        leftTask.join();
                        rightTask.join();
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                    }
                }

                merge(array, left, mid, right);
            }
        }

        // Classical merge sort for array with length below TRESHOLD
        private void mergeSort(int[] array, int left, int right) {
            if (left < right) {
                int mid = left + (right - left) / 2;

                mergeSort(array, left, mid);
                mergeSort(array, mid + 1, right);

                merge(array, left, mid, right);
            }
        }
    }

    private static void printArray(int[] array) {
        for (int i : array) {
            System.out.print(i + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        String filePath = "C:\\Users\\luisa\\Desktop\\SCS\\SCS_Project\\MS_input.txt";

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            // Read n
            String firstLine = br.readLine().trim();
            int n = Integer.parseInt(firstLine.split(":")[1].trim());

            // Read array
            String secondLine = br.readLine().trim();
            String[] arrayStrings = secondLine.split(":")[1].trim().split(" ");
            int[] array = new int[n];
            for (int i = 0; i < n; i++) {
                array[i] = Integer.parseInt(arrayStrings[i]);
            }

//            System.out.println("Number of elements: " + n);
//            System.out.print("Unsorted array: ");
//            printArray(array);

            long startTime = System.nanoTime();
            MergeSortTask task = new MergeSortTask(array, 0, n - 1);
            task.start();

            try {
                task.join();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                System.out.println("Merge sort interrupted");
            }

            long endTime = System.nanoTime();

            double durationInSeconds = (endTime - startTime) / 1_000_000_000.0;

            Thread[] threads = new Thread[n];
            long startContextSwitch = System.nanoTime();

            double totalThreadCreationTime = 0.0;

            for (int i = 0; i < n; i++) {
                final int threadId = i + 1;

                long threadStartTime = System.nanoTime();
                threads[i] = new WorkerThread(threadId);
                threads[i].start();
                long threadEndTime = System.nanoTime();

                totalThreadCreationTime += (threadEndTime - threadStartTime) / 1_000_000_000.0;
            }

            for (int i = 0; i < n; i++) {
                try {
                    threads[i].join();
                } catch (InterruptedException e) {
                    System.err.println("Thread " + i + " was interrupted.");
                }
            }

            long endContextSwitch = System.nanoTime();

            double avgThreadCreationTime = totalThreadCreationTime / n;
            double contextSwitchTime = (endContextSwitch - startContextSwitch) / 1_000_000_000.0;

            String outFilePath = "C:\\Users\\luisa\\Desktop\\SCS\\SCS_Project\\outputs.txt";
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(outFilePath, true))) {
                writer.write("\nParallel merge sort in Java results:\n");
                writer.write("\t(ms-java) Execution time: " + String.format(Locale.US,"%.6f", durationInSeconds) + " seconds.\n");
                writer.write("\t(ms-java) Average thread creation time: " + String.format(Locale.US,"%.6f", avgThreadCreationTime) + " seconds.\n");
                writer.write("\t(ms-java) Context switch time: " + String.format(Locale.US,"%.6f", contextSwitchTime) + " seconds.\n");
            } catch (IOException e) {
                System.out.println("Error writing to file: " + e.getMessage());
            }

//            System.out.print("Sorted array: ");
//            printArray(array);

        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
        } catch (NumberFormatException e) {
            System.out.println("Error parsing input: " + e.getMessage());
        }
    }
}
